framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,901 @@
+                } catch (IOException e) {
+                    ErrorReporter.errorAbort("Problem while reading file in: " + jaifPath + "."
+                            + " Exception message: " + e.getMessage());
+                }
+            }
+            scenes.put(jaifPath, scene);
+        }
+        return scenes.get(jaifPath);
+    }
+
+    /**
+     * Write all modified scenes into .jaif files.
+     * (Scenes are modified by the method {@link #updateAnnotationSetInScene}.)
+     */
+    public static void writeScenesToJaif() {
+        // Create .jaif files directory if it doesn't exist already.
+        File jaifDir = new File(JAIF_FILES_PATH);
+        if (!jaifDir.exists()) {
+            jaifDir.mkdirs();
+        }
+        // Write scenes into .jaif files.
+        for (String jaifPath : modifiedScenes) {
+            try {
+                AScene scene = scenes.get(jaifPath);
+                removeIgnoredAnnosFromScene(scene);
+                if (!scene.prune()) {
+                    // Only write non-empty scenes into .jaif files.
+                    IndexFileWriter.write(scene, new FileWriter(jaifPath));
+                }
+            } catch (IOException e) {
+                ErrorReporter.errorAbort("Problem while reading file in: " + jaifPath
+                        + ". Exception message: " + e.getMessage());
+            } catch (DefException e) {
+                ErrorReporter.errorAbort(e.getMessage());
+            }
+        }
+        modifiedScenes.clear();
+    }
+
+    /**
+     * Removes all annotations that should be ignored from an AScene.
+     * (See {@link #shouldIgnore}).
+     */
+    private static void removeIgnoredAnnosFromScene(AScene scene) {
+        for (AClass aclass : scene.classes.values()) {
+            for (AField field : aclass.fields.values()) {
+                removeIgnoredAnnosFromATypeElement(
+                        field.type, TypeUseLocation.FIELD);
+            }
+            for (AMethod method : aclass.methods.values()) {
+                // Return type
+                removeIgnoredAnnosFromATypeElement(
+                        method.returnType, TypeUseLocation.RETURN);
+                // Parameter type
+                for (AField param : method.parameters.values()) {
+                    removeIgnoredAnnosFromATypeElement(
+                            param.type, TypeUseLocation.PARAMETER);
+                }
+            }
+        }
+    }
+
+    /**
+     * Removes all annotations that should be ignored from an ATypeElement.
+     * (See {@link #shouldIgnore}).
+     */
+    private static void removeIgnoredAnnosFromATypeElement(ATypeElement typeEl,
+            TypeUseLocation loc) {
+        Set<Annotation> annosToRemove = new HashSet<>();
+        Set<String> annosToIgnoreForLocation = annosToIgnore.get(loc);
+        if (annosToIgnoreForLocation == null) return; // No annotations to ignore for that position.
```

Comment:
I find control like this hard to miss. I would make four lines out of this one line.